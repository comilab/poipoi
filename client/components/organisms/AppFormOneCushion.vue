<template lang="pug">
ValidationObserver(
  ref="form",
  v-slot="{ invalid }",
  slim
)
  form.max-w-screen-md.text-center.bg-white.rounded.mx-auto.p-8(
    @submit.prevent="onSubmit",
    name="one-cushion",
    aria-label="ワンクッション"
  )
    font-awesome-icon.fa-10x.text-red-700(
      :icon="icon"
    )

    .my-8
      div(v-if="post.rating === 'r18'") 18歳未満の方は以降の内容を閲覧することができません。
      div(v-else-if="post.rating === 'nsfw'") 閲覧に注意が必要な内容が含まれます。

      template(v-if="requirePassword")
        div 閲覧にはパスワードが必要です。
        AppInput.text-left.my-4(
          vid="password",
          label="パスワード",
          v-model="input.password",
          type="password",
          :rules="{ required: true }"
        )

    AppButton(
      type="submit",
      color="red",
      :icon="faCheck",
      :indeterminate="verifying",
      :disabled="invalid || verifying"
    )
      .inline-flex.flex-col.md_inline(
        v-if="post.rating === 'r18'"
      )
        span 18歳以上であることに
        span 同意した上で閲覧する
      span(v-else) 閲覧する
</template>

<script lang="ts">
import { defineComponent, reactive, toRefs, computed, ref } from 'nuxt-composition-api'
import { faExclamationTriangle, faLock, faCheck } from '@fortawesome/free-solid-svg-icons'

import Post from '~/models/Post'
import usePostsApi from '~/composables/use-posts-api'
import AppButton from '~/components/atoms/AppButton.vue'
import AppInput from '~/components/molecules/AppInput.vue'

export default defineComponent({
  components: {
    AppButton,
    AppInput
  },
  props: {
    post: {
      type: Post,
      required: true
    }
  },
  setup (props, { emit }) {
    const { post, verify, verifying } = usePostsApi(props.post)

    const form = ref<any>()

    const localState = reactive({
      input: {
        password: ''
      }
    })

    const icon = computed(() => {
      if (props.post.rating) {
        return faExclamationTriangle
      }
      return faLock
    })

    const requirePassword = computed(() => props.post.scope === 'password')

    const onSubmit = async () => {
      if (!requirePassword.value) {
        emit('passed', props.post)
      } else {
        try {
          await verify(localState.input)
          emit('passed', post.value)
        } catch (err) {
          form.value!.setErrors({
            password: 'パスワードが違います'
          })
        }
      }
    }

    return {
      form,
      verifying,
      ...toRefs(localState),
      icon,
      requirePassword,
      onSubmit,
      faCheck
    }
  }
})
</script>
