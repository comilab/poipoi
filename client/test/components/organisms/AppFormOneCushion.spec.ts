import { fireEvent, render, waitFor, within } from '@testing-library/vue'
import { mocked } from 'ts-jest/utils'
import { defineComponent, Ref, ref } from '@vue/composition-api'
import { faExclamationTriangle, faLock } from '@fortawesome/free-solid-svg-icons'

import AppFormOneCushion from '~/components/organisms/AppFormOneCushion.vue'
import usePostsApi from '~/composables/use-posts-api'
import postFactory from '~/test/factories/post'

jest.mock('~/composables/use-posts-api')
jest.useFakeTimers()

describe('components/organisms/AppFormOneCushion', () => {
  let options: any
  let usePostsApiMock: any

  beforeEach(() => {
    options = {
      props: {
        post: postFactory.build()
      },
      stubs: {
        FontAwesomeIcon: true
      }
    }
    usePostsApiMock = {
      verify: jest.fn().mockResolvedValue(null),
      verifying: ref(false)
    }
    mocked(usePostsApi).mockImplementation((post) => {
      return {
        ...usePostsApiMock,
        post: ref(post)
      }
    })
  })

  describe('props', () => {
    describe('post', () => {
      describe('rating = r18', () => {
        beforeEach(() => {
          options.props.post = postFactory.build({
            rating: 'r18',
            scope: 'public'
          })
        })

        test('renders correctly', () => {
          const { container } = render(AppFormOneCushion, options)

          expect(container).toMatchSnapshot()
        })

        test('faExclamationTriangleを表示', () => {
          let fontAwesomeIconProps: any
          options.stubs.AppButton = true
          options.stubs.FontAwesomeIcon = defineComponent({
            props: ['icon'],
            setup (props) {
              fontAwesomeIconProps = props
            },
            template: '<span />'
          })
          render(AppFormOneCushion, options)

          expect(fontAwesomeIconProps.icon).toStrictEqual(faExclamationTriangle)
        })
      })

      describe('rating = nsfw', () => {
        beforeEach(() => {
          options.props.post = postFactory.build({
            rating: 'nsfw',
            scope: 'public'
          })
        })

        test('renders correctly', () => {
          const { container } = render(AppFormOneCushion, options)

          expect(container).toMatchSnapshot()
        })

        test('faExclamationTriangleを表示', () => {
          let fontAwesomeIconProps: any
          options.stubs.AppButton = true
          options.stubs.FontAwesomeIcon = defineComponent({
            props: ['icon'],
            setup (props) {
              fontAwesomeIconProps = props
            },
            template: '<span />'
          })
          render(AppFormOneCushion, options)

          expect(fontAwesomeIconProps.icon).toStrictEqual(faExclamationTriangle)
        })
      })

      describe('scope = password', () => {
        beforeEach(() => {
          options.props.post = postFactory.build({
            rating: null,
            scope: 'password'
          })
        })

        test('renders correctly', () => {
          const { container } = render(AppFormOneCushion, options)

          expect(container).toMatchSnapshot()
        })

        test('faLockを表示', () => {
          let fontAwesomeIconProps: any
          options.stubs.AppButton = true
          options.stubs.FontAwesomeIcon = defineComponent({
            props: ['icon'],
            setup (props) {
              fontAwesomeIconProps = props
            },
            template: '<span />'
          })
          render(AppFormOneCushion, options)

          expect(fontAwesomeIconProps.icon).toStrictEqual(faLock)
        })

        test('パスワード入力欄を表示', () => {
          const { getByLabelText } = render(AppFormOneCushion, options)

          expect(getByLabelText('パスワード')).toBeInTheDocument()
        })
      })
    })
  })

  describe('form', () => {
    describe('inputs', () => {
      beforeEach(() => {
        options.props.post = postFactory.build({
          scope: 'password'
        })
      })

      describe('パスワード', () => {
        test('必須', async () => {
          const { getByLabelText, getByTestId } = render(AppFormOneCushion, options)
          await fireEvent.update(getByLabelText('パスワード'), '')
          jest.runAllTimers()

          await waitFor(() => {
            expect(getByTestId('input-パスワード')).toHaveClass('invalid')
          })
        })
      })
    })

    describe('submit', () => {
      let invalid: Ref<boolean>

      beforeEach(() => {
        invalid = ref(false)
        options.stubs.ValidationObserver = defineComponent({
          setup () {
            return { invalid }
          },
          template: '<div><slot :invalid="invalid" /></div>'
        })
      })

      test('入力エラーがあれば送信不可', () => {
        invalid.value = true
        const { getByRole } = render(AppFormOneCushion, options)

        expect(getByRole('button')).toHaveAttribute('disabled')
      })

      test('入力エラーがなければ送信可能', () => {
        const { getByRole } = render(AppFormOneCushion, options)

        expect(getByRole('button')).not.toHaveAttribute('disabled')
      })

      test('送信中は送信不可', () => {
        usePostsApiMock.verifying.value = true
        const { getByRole } = render(AppFormOneCushion, options)

        expect(getByRole('button')).toHaveAttribute('disabled')
      })

      test('送信中はindeterminate = true', async () => {
        let buttonProps: any
        options.stubs.AppButton = defineComponent({
          props: ['indeterminate'],
          setup (props) {
            buttonProps = props
          },
          template: '<button />'
        })
        render(AppFormOneCushion, options)
        expect(buttonProps.indeterminate).toBeFalsy()

        usePostsApiMock.verifying!.value = true
        await waitFor(() => {
          expect(buttonProps.indeterminate).toBeTruthy()
        })
      })
    })

    describe('onSubmit', () => {
      describe('scope = passwordでない場合', () => {
        beforeEach(() => {
          options.props.post = postFactory.build({
            scope: 'public'
          })
        })

        test('scope = passwordでない場合はすぐにpassedイベントを発火', async () => {
          const { getByRole, emitted } = render(AppFormOneCushion, options)
          await fireEvent.submit(getByRole('form'))

          expect(emitted().passed).toHaveLength(1)
          expect(emitted().passed[0]).toStrictEqual([options.props.post])
        })
      })

      describe('scope = passwordの場合', () => {
        beforeEach(() => {
          options.props.post = postFactory.build({
            scope: 'password'
          })
        })

        test('verifyを実行', async () => {
          const { getByLabelText, getByRole } = render(AppFormOneCushion, options)
          await fireEvent.update(getByLabelText('パスワード'), 'foo')
          await fireEvent.submit(getByRole('form'))

          expect(usePostsApiMock.verify).toBeCalledWith({
            password: 'foo'
          })
        })

        test('verifyに成功したらpassedを発火', async () => {
          const { getByRole, emitted } = render(AppFormOneCushion, options)
          await fireEvent.submit(getByRole('form'))

          await waitFor(() => {
            expect(emitted().passed).toHaveLength(1)
            expect(emitted().passed[0]).toStrictEqual([options.props.post])
          })
        })

        test('verifyに失敗したらエラーを表示', async () => {
          usePostsApiMock.verify.mockRejectedValue()
          const { getByRole, emitted, getByTestId } = render(AppFormOneCushion, options)
          await fireEvent.submit(getByRole('form'))

          await waitFor(() => {
            expect(emitted().passed).toBeFalsy()
            within(getByTestId('input-パスワード')).getByText('パスワードが違います')
          })
        })
      })
    })
  })
})
