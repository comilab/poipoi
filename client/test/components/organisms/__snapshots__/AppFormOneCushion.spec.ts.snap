// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/AppFormOneCushion props post rating = nsfw renders correctly 1`] = `
<div>
  <form
    aria-label="ワンクッション"
    class="max-w-screen-md text-center bg-white rounded mx-auto p-8"
    name="one-cushion"
  >
    <fontawesomeicon-stub
      class="fa-10x text-red-700"
      icon="[object Object]"
    />
    <div
      class="my-8"
    >
      <div>
        閲覧に注意が必要な内容が含まれます。
      </div>
      <!---->
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-red-700 text-white hover_bg-red-800 px-3 py-2 rounded"
      data-testid="button"
      type="submit"
    >
      <div>
        <fontawesomeicon-stub
          class="fa-fw mr-2"
          data-testid="icon"
          icon="[object Object]"
        />
      </div>
      <div
        class=""
        data-testid="text"
      >
        <span>
          閲覧する
        </span>
      </div>
    </button>
  </form>
</div>
`;

exports[`components/organisms/AppFormOneCushion props post rating = r18 renders correctly 1`] = `
<div>
  <form
    aria-label="ワンクッション"
    class="max-w-screen-md text-center bg-white rounded mx-auto p-8"
    name="one-cushion"
  >
    <fontawesomeicon-stub
      class="fa-10x text-red-700"
      icon="[object Object]"
    />
    <div
      class="my-8"
    >
      <div>
        18歳未満の方は以降の内容を閲覧することができません。
      </div>
      <!---->
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-red-700 text-white hover_bg-red-800 px-3 py-2 rounded"
      data-testid="button"
      type="submit"
    >
      <div>
        <fontawesomeicon-stub
          class="fa-fw mr-2"
          data-testid="icon"
          icon="[object Object]"
        />
      </div>
      <div
        class=""
        data-testid="text"
      >
        <div
          class="inline-flex flex-col md_inline"
        >
          <span>
            18歳以上であることに
          </span>
          <span>
            同意した上で閲覧する
          </span>
        </div>
      </div>
    </button>
  </form>
</div>
`;

exports[`components/organisms/AppFormOneCushion props post scope = password renders correctly 1`] = `
<div>
  <form
    aria-label="ワンクッション"
    class="max-w-screen-md text-center bg-white rounded mx-auto p-8"
    name="one-cushion"
  >
    <fontawesomeicon-stub
      class="fa-10x text-red-700"
      icon="[object Object]"
    />
    <div
      class="my-8"
    >
      <!---->
      <div>
        閲覧にはパスワードが必要です。
      </div>
      <div
        class="flex flex-col text-left my-4 untouched pristine"
        data-testid="input-パスワード"
      >
        <label
          class="text-sm font-bold py-2"
          data-testid="label"
          for="input-0"
        >
          パスワード
        </label>
        <input
          class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
          data-testid="input"
          id="input-0"
          name="パスワード"
          type="password"
        />
        <div
          class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
        >
          <div
            data-testid="errors"
          />
          <!---->
        </div>
      </div>
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-red-700 text-white hover_bg-red-800 px-3 py-2 rounded"
      data-testid="button"
      type="submit"
    >
      <div>
        <fontawesomeicon-stub
          class="fa-fw mr-2"
          data-testid="icon"
          icon="[object Object]"
        />
      </div>
      <div
        class=""
        data-testid="text"
      >
        <span>
          閲覧する
        </span>
      </div>
    </button>
  </form>
</div>
`;
